plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -65,86 +64,112 @@ private static ClusteringPushDownResult getClusteringKeysSet(List<CassandraColum
                         }
-                        else if (!rangeConjuncts.isEmpty()) {
-                            predicate = Joiner.on(" AND ").join(rangeConjuncts);
+                        if (ranges.getOrderedRanges().stream().allMatch(Range::isSingleValue)) {
+                            if (isInStatementNotAllowed(clusteringColumns, cassandraVersion, currentlyProcessedClusteringColumn)) {
+                                return null;
+                            }
+
+                            String inValues = ranges.getOrderedRanges().stream()
+                                    .map(range -> columnHandle.getCassandraType().toCqlLiteral(range.getSingleValue()))
+                                    .collect(joining(","));
+                            fullyPushedDomains.put(columnHandle, domain);
+                            return CassandraCqlUtils.validColumnName(columnHandle.getName()) + "IN (" + inValues + ")";
                         }
-                        return predicate;
+                        return translateRangeIntoCQL(columnHandle, ranges.getSpan());
                     }, discreteValues -> {
                         if (discreteValues.isInclusive()) {
-                            ImmutableList.Builder<Object> discreteValuesList = ImmutableList.builder();
-                            for (Object discreteValue : discreteValues.getValues()) {
-                                discreteValuesList.add(columnHandle.getCassandraType().toCqlLiteral(discreteValue));
+                            if (discreteValues.getValuesCount() == 0) {
+                                return null;
+                            }
+                            if (discreteValues.getValuesCount() == 1) {
+                                fullyPushedDomains.put(columnHandle, domain);
+                                return format("%s = %s",
+                                        CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                                        discreteValues.getValues().stream().findFirst()
+                                                .orElseThrow());
+                            }
+                            if (isInStatementNotAllowed(clusteringColumns, cassandraVersion, currentlyProcessedClusteringColumn)) {
+                                return null;
                             }
-                            String predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
-                                    + Joiner.on(",").join(discreteValuesList.build()) + ")";
-                            return predicate;
+
+                            fullyPushedDomains.put(columnHandle, domain);
+
+                            return discreteValues.getValues().stream()
+                                    .map(columnHandle.getCassandraType()::toCqlLiteral)
+                                    .reduce((first, second) -> first + "," + second)
```

Comment:
is this `O(n²)` ?

use `.collect( [Collectors.] joining(",", "IN(", ")") )`

Reply:
It's `O(n) `. I can change it to collect